lib/resty/core/regex.lua

```diff
@@ -478,6 +479,84 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
     return re_match_helper(subj, regex, opts, ctx, false, nil, nth)
 end
 
+local function destroy_re_gmatch_iterator(iterator)
+    if not iterator._compile_once then
+        destroy_compiled_regex(iterator._compiled)
+    end
+    iterator._pos = nil
+end
+
+local function iterate_re_gmatch(self)
+    local compiled = self._compiled
+    local subj = self._subj
+    local subj_len = self._subj_len
+    local flags = self._flags
+    local pos = self._pos
+
+    if not pos then
+        -- The iterator is exhausted.
+        return nil
+    end
+
+    local rc = C.ngx_http_lua_ffi_exec_regex(compiled, flags, subj,
+                                             subj_len, pos)
+
+    if rc == PCRE_ERROR_NOMATCH then
+        destroy_re_gmatch_iterator(self)
+        return nil
+    end
+
+    if rc < 0 then
+        destroy_re_gmatch_iterator(self)
+        return nil, "pcre_exec() failed: " .. rc
+    end
+
+    if rc == 0 then
+        if band(flags, FLAG_DFA) == 0 then
+            self._pos = nil
+            return nil, "capture size too small"
+        end
+
+        rc = 1
+    end
+
+    local cp_pos = compiled.captures[1]
+    if cp_pos == compiled.captures[0] then
+        cp_pos = cp_pos + 1
+        if cp_pos > subj_len then
+            destroy_re_gmatch_iterator(self)
+            return collect_captures(compiled, rc, subj, flags)
+        end
+    end
+    self._pos = cp_pos
+    return collect_captures(compiled, rc, subj, flags)
+end
+
+
+do
+    local re_gmatch_iterator_mt = { __call = iterate_re_gmatch }
+
+    function ngx.re.gmatch(subj, regex, opts)
+        subj  = tostring(subj)
+
+        local compiled, compile_once, flags = re_match_compile(regex, opts)
+        if compiled == nil then
+            -- compiled_once holds the error string
+            return nil, compile_once
+        end
+
+        local re_gmatch_iterator = new_tab(0, 6)
+        re_gmatch_iterator._compiled = compiled
+        re_gmatch_iterator._compile_once = compile_once
+        re_gmatch_iterator._subj = subj
+        re_gmatch_iterator._subj_len = #subj
+        re_gmatch_iterator._flags = flags
+        re_gmatch_iterator._pos = 0
```

Comment:
The lines above look too verbose and cumbersome. I think it's nicer and also faster to use the following form instead:

```
local re_match_iterator = {
    _compiled = compiled,
    _compie_once = compile_once,
    _subj = subj,
    _subj_len = #subj,
    _flags = flags,
    _pos = 0,
}
```